### PR TITLE
Ground track analysis

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: astronomy/orbit_ground_track.hpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+using physics::RotatingBody;
+using quantities::Angle;
+using quantities::AngularFrequency;
+using quantities::Time;
+
+class OrbitGroundTrack {
+ public:
+  struct MeanSun {
+    Instant epoch;
+    // This mean longitude of the sun is with respect the axes of
```
Comment (pleroy) [inline astronomy/orbit_ground_track.hpp]: to

File: astronomy/orbit_ground_track.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    Angle mean_longitude_at_epoch;
+    // This year is the period of the above mean longitude; which kind of year
+    // it is depends on the definition of |PrimaryCentred|.  If |PrimaryCentred|
+    // has ICRS axes, it is the sidereal year; if |PrimaryCentred| is the
+    // reference frame of the equator & equinox of the date, it is the tropical
+    // year.
+    Time year;
+  };
+
+  template<typename PrimaryCentred, typename Inertial>
```
Comment (pleroy) [inline astronomy/orbit_ground_track.hpp]: Comment this function.

File: astronomy/orbit_ground_track_body.hpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+// GCRS, this is the right ascension (with respect to the mean equinox of
+// J2000).
+template<typename PrimaryCentred>
+Angle CelestialLongitude(Position<PrimaryCentred> const& q) {
+  // TODO(egg): |ToSpherical| is wasteful, as we discard the latitude.
+  return (q - PrimaryCentred::origin).coordinates().ToSpherical().longitude;
+}
+
+// The resulting angle is not normalized.
+template<typename Iterator, typename Inertial>
```
Comment (pleroy) [inline astronomy/orbit_ground_track_body.hpp]: Why is this taking an `Iterator` (which could be anything) and not a frame (where the iterator would be `DiscreteTrajectory<Frame>::Iterator`)?

File: astronomy/orbit_ground_track_body.hpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+template<typename PrimaryCentred, typename Inertial>
+Interval<Angle> ReducedLongitudesOfEquatorialCrossings(
+    DiscreteTrajectory<PrimaryCentred> const& nodes,
+    RotatingBody<Inertial> const& primary,
+    OrbitRecurrence const& nominal_recurrence,
+    std::optional<Angle>& initial_offset) {
+  Interval<Angle> reduced_longitudes;
+  int n = 0;
```
Comment (pleroy) [inline astronomy/orbit_ground_track_body.hpp]: This is only used to compute the equatorial shift, you could do that by addition through the loop, that would be more elegant.

File: geometry/interval.hpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+namespace principia {
+namespace geometry {
+namespace internal_interval {
+
+using quantities::Difference;
+using quantities::Infinity;
+
+// Represents the interval [min, max]; T must be an ordered affine space.
+template<typename T>
+struct Interval {
```
Comment (pleroy) [inline geometry/interval.hpp]: I would prefer it if this was a class and `min` and `max` were selectors.  You should really not shoot directly at `min` and `max`.

File: astronomy/orbit_analysis_test.cpp [hunk 6]
```diff
@@ -156,18 +171,64 @@ class OrbitAnalysisTest : public ::testing::Test {
...
                                            /*max_abs_Cᴛₒ=*/100);
-    return {elements, recurrence};
+    // Since our ITRS-to-ICRS conversion disregards the precession of the
+    // equinoxes, it is not completely clear which year we should be dealing
+    // with here.  Given that the SP3 files have data in the ITRS (whose equator
+    // is the equator of date), and that we map that to the ICRS equator, our
+    // IRCS here is more like an equator-of-date frame (although it is not clear
+    // whether its x axis resembles the equinox of the date), so we pick
+    // the tropical year.
+    // We use values based Newcomb's formula for the mean longitude of the sun,
```
Comment (pleroy) [inline astronomy/orbit_analysis_test.cpp]: on

File: astronomy/orbit_ground_track_body.hpp (hunk 3 above)
Comment (eggrobin) [reply astronomy/orbit_ground_track_body.hpp]: Deduction.

File: geometry/interval.hpp (hunk 5 above)
Comment (eggrobin) [reply geometry/interval.hpp]: Maybe later.

**Comment by eggrobin** [comment]
retest this please

**Pull request merged**